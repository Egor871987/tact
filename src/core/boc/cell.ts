/**
 * Copyright (c) Whales Corp.
 * All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import inspectSymbol from "symbol.inspect";
import { BitString } from "@/core/boc/bit-string";
import { CellType } from "@/core/boc/cell-type";
import { Slice } from "@/core/boc/slice";
import type { LevelMask } from "@/core/boc/cell/level-mask";
import { resolveExotic } from "@/core/boc/cell/resolve-exotic";
import { wonderCalculator } from "@/core/boc/cell/wonder-calculator";
import { deserializeBoc, serializeBoc } from "@/core/boc/cell/serialization";
import { BitReader } from "@/core/boc/bit-reader";
import { beginCell } from "@/core/boc/builder";

/**
 * Cell as described in TVM spec
 */
export class Cell {
    static readonly EMPTY = new Cell();

    /**
     * Deserialize cells from BOC
     * @param src source buffer
     * @returns array of cells
     */
    static fromBoc(src: Buffer) {
        return deserializeBoc(src);
    }

    /**
     * Helper function that deserializes a single cell from BOC in base64
     * @param src source string
     */
    static fromBase64(src: string): Cell {
        const parsed = Cell.fromBoc(Buffer.from(src, "base64"));
        const cell = parsed[0];
        if (parsed.length !== 1 || typeof cell === "undefined") {
            throw new Error("Deserialized more than one cell");
        }
        return cell;
    }

    /**
     * Helper function that deserializes a single cell from BOC in hex
     * @param src source string
     */
    static fromHex(src: string): Cell {
        const parsed = Cell.fromBoc(Buffer.from(src, "hex"));
        const cell = parsed[0];
        if (parsed.length !== 1 || typeof cell === "undefined") {
            throw new Error("Deserialized more than one cell");
        }
        return cell;
    }

    // Public properties
    readonly type: CellType;
    readonly bits: BitString;
    readonly refs: Cell[];
    readonly mask: LevelMask;

    // Level and depth information
    private _hashes: Buffer[] = [];
    private _depths: number[] = [];

    constructor(opts?: { exotic?: boolean; bits?: BitString; refs?: Cell[] }) {
        // Resolve bits
        let bits = BitString.EMPTY;
        if (opts?.bits) {
            bits = opts.bits;
        }

        // Resolve refs
        let refs: Cell[] = [];
        if (opts?.refs) {
            refs = [...opts.refs];
        }

        // Resolve type
        let hashes: Buffer[];
        let depths: number[];
        let mask: LevelMask;
        let type = CellType.Ordinary;
        if (opts?.exotic) {
            // Resolve exotic cell
            const resolved = resolveExotic(bits, refs);

            // Perform wonders
            const wonders = wonderCalculator(resolved.type, bits, refs);

            // Copy results
            mask = wonders.mask;
            depths = wonders.depths;
            hashes = wonders.hashes;
            type = resolved.type;
        } else {
            // Check correctness
            if (refs.length > 4) {
                throw new Error("Invalid number of references");
            }
            if (bits.length > 1023) {
                throw new Error(`Bits overflow: ${bits.length} > 1023`);
            }

            // Perform wonders
            const wonders = wonderCalculator(CellType.Ordinary, bits, refs);

            // Copy results
            mask = wonders.mask;
            depths = wonders.depths;
            hashes = wonders.hashes;
            type = CellType.Ordinary;
        }

        // Set fields
        this.type = type;
        this.bits = bits;
        this.refs = refs;
        this.mask = mask;
        this._depths = depths;
        this._hashes = hashes;

        Object.freeze(this);
        Object.freeze(this.refs);
        Object.freeze(this.bits);
        Object.freeze(this.mask);
        Object.freeze(this._depths);
        Object.freeze(this._hashes);
    }

    /**
     * Check if cell is exotic
     */
    get isExotic() {
        return this.type !== CellType.Ordinary;
    }

    /**
     * Begin cell parsing
     * @returns a new slice
     */
    beginParse = (allowExotic: boolean = false) => {
        if (this.isExotic && !allowExotic) {
            throw new Error("Exotic cells cannot be parsed");
        }
        return new Slice(new BitReader(this.bits), this.refs);
    };

    /**
     * Get cell hash
     * @param level level
     * @returns cell hash
     */
    hash = (level: number = 3): Buffer => {
        const hashId = Math.min(this._hashes.length - 1, level);
        const hash = this._hashes[hashId];
        if (typeof hash === "undefined") {
            throw new Error("Invalid level");
        }
        return hash;
    };

    /**
     * Get cell depth
     * @param level level
     * @returns cell depth
     */
    depth = (level: number = 3): number => {
        const hashId = Math.min(this._depths.length - 1, level);
        const hash = this._depths[hashId];
        if (typeof hash === "undefined") {
            throw new Error("Invalid level");
        }
        return hash;
    };

    /**
     * Get cell level
     * @returns cell level
     */
    level = (): number => {
        return this.mask.level;
    };

    /**
     * Checks cell to be equal to another cell
     * @param other other cell
     * @returns true if cells are equal
     */
    equals = (other: Cell): boolean => {
        return this.hash().equals(other.hash());
    };

    /**
     * Serializes cell to BOC
     * @param opts options
     */
    toBoc(opts?: {
        idx?: boolean | null | undefined;
        crc32?: boolean | null | undefined;
    }): Buffer {
        const idx =
            opts && opts.idx !== null && opts.idx !== undefined
                ? opts.idx
                : false;
        const crc32 =
            opts && opts.crc32 !== null && opts.crc32 !== undefined
                ? opts.crc32
                : true;
        return serializeBoc(this, { idx, crc32 });
    }

    /**
     * Format cell to string
     * @param indent indentation
     * @returns string representation
     */
    toString(indent?: string): string {
        const id = indent ?? "";
        let t = "x";
        if (this.isExotic) {
            if (this.type === CellType.MerkleProof) {
                t = "p";
            } else if (this.type === CellType.MerkleUpdate) {
                t = "u";
            } else if (this.type === CellType.PrunedBranch) {
                t = "p";
            }
        }
        let s =
            id + (this.isExotic ? t : "x") + "{" + this.bits.toString() + "}";
        for (const i of this.refs) {
            s += "\n" + i.toString(id + " ");
        }
        return s;
    }

    /**
     * Convert cell to slice
     * @returns slice
     */
    asSlice() {
        return this.beginParse();
    }

    /**
     * Convert cell to a builder that has this cell stored
     * @returns builder
     */
    asBuilder() {
        return beginCell().storeSlice(this.asSlice());
    }

    [inspectSymbol] = () => this.toString();
}
