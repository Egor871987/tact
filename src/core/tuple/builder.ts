/**
 * Copyright (c) Whales Corp.
 * All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import type { Address } from "../address/address";
import { beginCell } from "../boc/builder";
import { Cell } from "../boc/cell";
import { Slice } from "../boc/slice";
import type { Maybe } from "../utils/maybe";
import type { TupleItem } from "./tuple";

export class TupleBuilder {
    private _tuple: TupleItem[] = [];

    writeNumber(v?: Maybe<bigint | number>) {
        if (v === null || v === undefined) {
            this._tuple.push({ type: "null" });
        } else {
            this._tuple.push({ type: "int", value: BigInt(v) });
        }
    }

    writeBoolean(v?: Maybe<boolean>) {
        if (v === null || v === undefined) {
            this._tuple.push({ type: "null" });
        } else {
            this._tuple.push({ type: "int", value: v ? -1n : 0n });
        }
    }

    writeBuffer(v?: Maybe<Buffer | null | undefined>) {
        if (v === null || v === undefined) {
            this._tuple.push({ type: "null" });
        } else {
            this._tuple.push({
                type: "slice",
                cell: beginCell().storeBuffer(v).endCell(),
            });
        }
    }

    writeString(v?: Maybe<string>) {
        if (v === null || v === undefined) {
            this._tuple.push({ type: "null" });
        } else {
            this._tuple.push({
                type: "slice",
                cell: beginCell().storeStringTail(v).endCell(),
            });
        }
    }

    writeCell(v?: Maybe<Cell | Slice>) {
        if (v === null || v === undefined) {
            this._tuple.push({ type: "null" });
        } else {
            if (v instanceof Cell) {
                this._tuple.push({ type: "cell", cell: v });
            } else if (v instanceof Slice) {
                this._tuple.push({ type: "cell", cell: v.asCell() });
            }
        }
    }

    writeSlice(v?: Maybe<Cell | Slice>) {
        if (v === null || v === undefined) {
            this._tuple.push({ type: "null" });
        } else {
            if (v instanceof Cell) {
                this._tuple.push({ type: "slice", cell: v });
            } else if (v instanceof Slice) {
                this._tuple.push({ type: "slice", cell: v.asCell() });
            }
        }
    }

    writeBuilder(v?: Maybe<Cell | Slice>) {
        if (v === null || v === undefined) {
            this._tuple.push({ type: "null" });
        } else {
            if (v instanceof Cell) {
                this._tuple.push({ type: "builder", cell: v });
            } else if (v instanceof Slice) {
                this._tuple.push({ type: "builder", cell: v.asCell() });
            }
        }
    }

    writeTuple(v?: Maybe<TupleItem[]>) {
        if (v === null || v === undefined) {
            this._tuple.push({ type: "null" });
        } else {
            this._tuple.push({ type: "tuple", items: v });
        }
    }

    writeAddress(v?: Maybe<Address>) {
        if (v === null || v === undefined) {
            this._tuple.push({ type: "null" });
        } else {
            this._tuple.push({
                type: "slice",
                cell: beginCell().storeAddress(v).endCell(),
            });
        }
    }

    build() {
        return [...this._tuple];
    }
}
