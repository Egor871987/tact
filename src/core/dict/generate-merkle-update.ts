import { beginCell } from "@/core/boc/builder";
import type { Cell } from "@/core/boc/cell";
import type {
    DictionaryKeyTypes,
    Dictionary,
    DictionaryKey,
} from "@/core/dict/dictionary";
import { generateMerkleProof } from "@/core/dict/generate-merkle-proof";

function convertToMerkleUpdate(c1: Cell, c2: Cell): Cell {
    return beginCell()
        .storeUint(4, 8)
        .storeBuffer(c1.hash(0))
        .storeBuffer(c2.hash(0))
        .storeUint(c1.depth(0), 16)
        .storeUint(c2.depth(0), 16)
        .storeRef(c1)
        .storeRef(c2)
        .endCell({ exotic: true });
}

export function generateMerkleUpdate<K extends DictionaryKeyTypes, V>(
    dict: Dictionary<K, V>,
    key: K,
    keyObject: DictionaryKey<K>,
    newValue: V,
): Cell {
    const oldProof = generateMerkleProof(dict, [key], keyObject).refs[0];
    dict.set(key, newValue);
    const newProof = generateMerkleProof(dict, [key], keyObject).refs[0];
    if (typeof oldProof === 'undefined' || typeof newProof === 'undefined') {
        throw new Error("Bug");
    }
    return convertToMerkleUpdate(oldProof, newProof);
}
