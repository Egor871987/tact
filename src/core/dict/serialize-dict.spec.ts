/**
 * Copyright (c) Whales Corp.
 * All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { beginCell } from "@/core/boc/builder";
import { serializeDict } from "@/core/dict/serialize-dict";

describe("serializeDict", () => {
    it("should build prefix tree", () => {
        // From docs
        const map: Map<bigint, bigint> = new Map();
        map.set(13n, 169n);
        map.set(17n, 289n);
        map.set(239n, 57121n);

        // Test serialization
        const builder = beginCell();
        serializeDict(map, 16, (src, cell) => cell.storeUint(src, 16), builder);
        const root = builder.endCell();
        expect(root).toMatchSnapshot();
    });
});
