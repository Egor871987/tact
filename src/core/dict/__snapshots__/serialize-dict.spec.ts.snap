// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`serializeDict should build prefix tree 1`] = `
Cell {
  "_depths": [
    2,
    2,
    2,
    2,
  ],
  "_hashes": [
    Buffer{c8c0ca7071eabf18a71adcbb398d1d2164b1378b9ae70c00510049fb865aec6a},
    Buffer{c8c0ca7071eabf18a71adcbb398d1d2164b1378b9ae70c00510049fb865aec6a},
    Buffer{c8c0ca7071eabf18a71adcbb398d1d2164b1378b9ae70c00510049fb865aec6a},
    Buffer{c8c0ca7071eabf18a71adcbb398d1d2164b1378b9ae70c00510049fb865aec6a},
  ],
  "beginParse": [Function],
  "bits": BitString {
    "_data": Buffer{c800000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000},
    "_length": 8,
    "_offset": 0,
    Symbol(nodejs.util.inspect.custom): [Function],
  },
  "depth": [Function],
  "equals": [Function],
  "hash": [Function],
  "level": [Function],
  "mask": LevelMask {
    "_hashCount": 1,
    "_hashIndex": 0,
    "_mask": 0,
  },
  "refs": [
    Cell {
      "_depths": [
        1,
        1,
        1,
        1,
      ],
      "_hashes": [
        Buffer{c615ca8b15809e9db17119e66b7dc324b5bbd248c8caf2fe6356676e7d6b800e},
        Buffer{c615ca8b15809e9db17119e66b7dc324b5bbd248c8caf2fe6356676e7d6b800e},
        Buffer{c615ca8b15809e9db17119e66b7dc324b5bbd248c8caf2fe6356676e7d6b800e},
        Buffer{c615ca8b15809e9db17119e66b7dc324b5bbd248c8caf2fe6356676e7d6b800e},
      ],
      "beginParse": [Function],
      "bits": BitString {
        "_data": Buffer{6000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000},
        "_length": 6,
        "_offset": 0,
        Symbol(nodejs.util.inspect.custom): [Function],
      },
      "depth": [Function],
      "equals": [Function],
      "hash": [Function],
      "level": [Function],
      "mask": LevelMask {
        "_hashCount": 1,
        "_hashIndex": 0,
        "_mask": 0,
      },
      "refs": [
        Cell {
          "_depths": [
            0,
            0,
            0,
            0,
          ],
          "_hashes": [
            Buffer{c5cf368cde29d296d3d448d6140d3c2cf825e755ddd52bb3b8bd97c22adc9c77},
            Buffer{c5cf368cde29d296d3d448d6140d3c2cf825e755ddd52bb3b8bd97c22adc9c77},
            Buffer{c5cf368cde29d296d3d448d6140d3c2cf825e755ddd52bb3b8bd97c22adc9c77},
            Buffer{c5cf368cde29d296d3d448d6140d3c2cf825e755ddd52bb3b8bd97c22adc9c77},
          ],
          "beginParse": [Function],
          "bits": BitString {
            "_data": Buffer{a680548000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000},
            "_length": 25,
            "_offset": 0,
            Symbol(nodejs.util.inspect.custom): [Function],
          },
          "depth": [Function],
          "equals": [Function],
          "hash": [Function],
          "level": [Function],
          "mask": LevelMask {
            "_hashCount": 1,
            "_hashIndex": 0,
            "_mask": 0,
          },
          "refs": [],
          "type": -1,
          Symbol(nodejs.util.inspect.custom): [Function],
        },
        Cell {
          "_depths": [
            0,
            0,
            0,
            0,
          ],
          "_hashes": [
            Buffer{11ed2c52db114c9b013a694072ce0e6fbc0c0567e8ac53599c88e76ae51e3bbd},
            Buffer{11ed2c52db114c9b013a694072ce0e6fbc0c0567e8ac53599c88e76ae51e3bbd},
            Buffer{11ed2c52db114c9b013a694072ce0e6fbc0c0567e8ac53599c88e76ae51e3bbd},
            Buffer{11ed2c52db114c9b013a694072ce0e6fbc0c0567e8ac53599c88e76ae51e3bbd},
          ],
          "beginParse": [Function],
          "bits": BitString {
            "_data": Buffer{a080908000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000},
            "_length": 25,
            "_offset": 0,
            Symbol(nodejs.util.inspect.custom): [Function],
          },
          "depth": [Function],
          "equals": [Function],
          "hash": [Function],
          "level": [Function],
          "mask": LevelMask {
            "_hashCount": 1,
            "_hashIndex": 0,
            "_mask": 0,
          },
          "refs": [],
          "type": -1,
          Symbol(nodejs.util.inspect.custom): [Function],
        },
      ],
      "type": -1,
      Symbol(nodejs.util.inspect.custom): [Function],
    },
    Cell {
      "_depths": [
        0,
        0,
        0,
        0,
      ],
      "_hashes": [
        Buffer{56a96869dacf6909c81210f7125275349ec253a143c4b2c001c7f8d68332842a},
        Buffer{56a96869dacf6909c81210f7125275349ec253a143c4b2c001c7f8d68332842a},
        Buffer{56a96869dacf6909c81210f7125275349ec253a143c4b2c001c7f8d68332842a},
        Buffer{56a96869dacf6909c81210f7125275349ec253a143c4b2c001c7f8d68332842a},
      ],
      "beginParse": [Function],
      "bits": BitString {
        "_data": Buffer{befdf21000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000},
        "_length": 28,
        "_offset": 0,
        Symbol(nodejs.util.inspect.custom): [Function],
      },
      "depth": [Function],
      "equals": [Function],
      "hash": [Function],
      "level": [Function],
      "mask": LevelMask {
        "_hashCount": 1,
        "_hashIndex": 0,
        "_mask": 0,
      },
      "refs": [],
      "type": -1,
      Symbol(nodejs.util.inspect.custom): [Function],
    },
  ],
  "type": -1,
  Symbol(nodejs.util.inspect.custom): [Function],
}
`;
